warning: use of old-style cast to ‘void*’ [-Wold-style-cast]

Thank you for taking the time to submit an issue!

## Background information

### What version of Open MPI are you using? (e.g., v1.10.3, v2.1.0, git branch name and hash, etc.)
openmpi-4.0.0


### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)
Compiled from a source/distribution tarball.


### Please describe the system on which you are running

* Operating system/version: Ubuntu 18.10
* Computer hardware: x86_64 arch
* Network type: 

-----------------------------

## Details of the problem
This macro in mpi.h produces so many warnings about old-style-cast in any code using OpenMPI, e.g. OpenFOAM opensource package.

    #if !OMPI_BUILDING
    #define OMPI_PREDEFINED_GLOBAL(type, global) ((type) ((void *) &(global)))
    #else
    #define OMPI_PREDEFINED_GLOBAL(type, global) ((type) &(global))
    #endif

the warnings:

    In file included from db/IOstreams/Pstreams/OPwrite.C:29:
    db/IOstreams/Pstreams/OPwrite.C: In static member function ‘static bool Foam::OPstream::write(Foam::Pstream::commsTypes, int, const char*, std::streamsize, int, Foam::label)’:
    /home/shadowfax/foam/foam-extend-4.1/ThirdParty/packages/openmpi-4.0.0/platforms/linux64GccDPOpt/include/mpi.h:329:72: warning: use of old-style cast to ‘void*’ [-Wold-style-cast]
    #define OMPI_PREDEFINED_GLOBAL(type, global) ((type) ((void *) &(global)))
                                                                        ^
    /home/shadowfax/foam/foam-extend-4.1/ThirdParty/packages/openmpi-4.0.0/platforms/linux64GccDPOpt/include/mpi.h:1049:18: note: in expansion of macro ‘OMPI_PREDEFINED_GLOBAL’
    #define MPI_BYTE OMPI_PREDEFINED_GLOBAL(MPI_Datatype, ompi_mpi_byte)
                  ^~~~~~~~~~~~~~~~~~~~~~
    db/IOstreams/Pstreams/OPwrite.C:101:13: note: in expansion of macro ‘MPI_BYTE’
             MPI_BYTE,

The OpenFOAM package use openmpi to manage parallel operations, and it includes mpi.h from openmpi package anywhere it needs. The source code of OpenFOAM needs to be compiled with C++ compiler and that is why anywhere it includes and uses mpi.h function, the c++ compiler warns about the usage of old-style-cast.

yes, it should be static_cast<void*> instead of (void*).
I think every modern c++ compiler should support that because it is not new feature...

I have already tried to recompile openmpi-4.0.0 with the following modification, however it does not compile with the "mpi.h:329:55: error: 'static_cast' undeclared (first use in this function)" error.

    #define OMPI_PREDEFINED_GLOBAL(type, global) ((type) (static_cast<void *>( &(global))))

--- a/ompi/include/mpi.h.in
+++ b/ompi/include/mpi.h.in
@@ -9,7 +9,7 @@
  *                         University of Stuttgart.  All rights reserved.
  * Copyright (c) 2004-2005 The Regents of the University of California.
  *                         All rights reserved.
- * Copyright (c) 2007-2018 Cisco Systems, Inc.  All rights reserved
+ * Copyright (c) 2007-2019 Cisco Systems, Inc.  All rights reserved
  * Copyright (c) 2008-2009 Sun Microsystems, Inc.  All rights reserved.
  * Copyright (c) 2009-2012 Oak Rigde National Laboratory.  All rights reserved.
  * Copyright (c) 2011      Sandia National Laboratories. All rights reserved.
@@ -325,7 +325,11 @@
  * when building OMPI).
  */
 #if !OMPI_BUILDING
+#if defined(c_plusplus) || defined(__cplusplus)
+#define OMPI_PREDEFINED_GLOBAL(type, global) (static_cast<type> (static_cast<void *> (&(global))))
+#else
 #define OMPI_PREDEFINED_GLOBAL(type, global) ((type) ((void *) &(global)))
+#endif
 #else
 #define OMPI_PREDEFINED_GLOBAL(type, global) ((type) &(global))
 #endif
@@ -758,8 +762,13 @@ enum {
  */
 #define MPI_INFO_ENV OMPI_PREDEFINED_GLOBAL(MPI_Info, ompi_mpi_info_env)
 
+#if defined(c_plusplus) || defined(__cplusplus)
+#define MPI_STATUS_IGNORE (static_cast<MPI_Status *> (0))
+#define MPI_STATUSES_IGNORE (static_cast<MPI_Status *> (0))
+#else
 #define MPI_STATUS_IGNORE ((MPI_Status *) 0)
 #define MPI_STATUSES_IGNORE ((MPI_Status *) 0)
+#endif
 
 /*
  * Special MPI_T handles

